### ISSUE TITLE
Forwarding a RenderFragment through a component hierarchy.

### ISSUE THREAD

**Issue by JHBonarius** [dotnet/aspnetcore#30667]
### Describe the bug
(This is a new attempt to describe one issue in https://github.com/dotnet/aspnetcore/issues/30534)

I'm trying to pass a RenderFragment through a component hierarchy. I'll give a simple example:
```c#
@*Baz.razor*@
@if (Bar is null)
{
    <div>nobar</div>
}
else
{
    <div>Render bar: @Bar</div>
}

@code {
    [Parameter]
    public RenderFragment Bar { get; set; }
}
```
I can use this
```xml
<Baz />
```
which renders "nobar", and
```xml
<Baz><Bar>bar</Bar></Baz>
```
which renders "Render bar: bar".

But I want to have a hierarchy of components and forward parameters. Thus I add a component `Foo`
```c#
@*Foo.razor*@
<Baz Bar=@Bar />

@code {
    [Parameter]
    public RenderFragment Bar { get; set; }
}
```

So now again I can use
```xml
<Foo><Bar>bar</Bar></Foo>
```
which renders "Render bar: bar". But now when I do
```xml
<Foo />
```
I get the exception "Delegate to an instance method cannot have null 'this'."

This is a bug IMHO.

I've seemed to narrow this down to the way the Razor source generator translates this. The generated code for `Foo.razor.g.cs` looks like (simplified):
```c#
protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder)
{
    __builder.OpenComponent<BlazorServerDefault.Pages.Baz>(0);
    __builder.AddAttribute(1, "Bar", new Microsoft.AspNetCore.Components.RenderFragment(Bar));
    __builder.CloseComponent();
}
```
The `new Microsoft.AspNetCore.Components.RenderFragment(Bar)` throws the exception when `Bar` is `null`. 

I think I found the cause for this being generated, explained here: https://github.com/dotnet/aspnetcore/issues/30534#issuecomment-789737363.

I think it should generate (custom simplified code)
```c#
protected override void BuildRenderTree(RenderTreeBuilder builder)
{
    builder.OpenComponent<Baz>(0);
    builder.AddAttribute(1, "Bar", Bar);
    builder.CloseComponent();
}
```
Which does work, in both cases.

### To Reproduce
See Code above

### Exceptions (if any)
`ArgumentException: Delegate to an instance method cannot have null 'this'.`

### Further technical details
- ASP.NET Core version: 5
- Include the output of `dotnet --info` : See https://github.com/dotnet/aspnetcore/issues/30534
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version: VS 2019 16

**Comment by SteveSandersonMS** [comment]
Thanks for reporting this, @JHBonarius. It looks to me like it would be enough to have #30663 to track this issue, and close this one as a duplicate. In both cases the underlying issue is that the Razor compiler is wrapping `new RenderFragment(...)` around the supplied value.

Do you agree? If you think this issue tracks something different from #30663, could you please clarify how it's different?

**Comment by JHBonarius** [comment]
@SteveSandersonMS Yeah, ok, I had a single previous topic #30534 which had the two combined and you said it was too hard to follow, so I split them up. If it is clear for you now, then I will close it.
